### ISSUE TITLE
Haproxy 2.3.2 tcp-check 'send' and 'expect'

### ISSUE THREAD

**Issue by mhclaudiu** [haproxy/haproxy#1027]
## Detailed description of the problem

Good day, on 2.3.2 version of haproxy there is a small issue sending commands and matching responses.
For example:
```
option tcp-check
tcp-check connect port 11212
tcp-check send stats\r\n
tcp-check expect string STAT\ pid 
or the regex one tcp-check expect rstring ^STAT\ pid\ (.*[\d]).*$
```
This will send 'stats' to a memcached server, and the response it will be in this form:
```
STAT pid 366
STAT uptime 64808
```

This is working perfectly fine on 2.1.4 version, but not on 2.3.2 on which servers are considered down:
```
Layer7 timeout, info: " at step 4 of tcp-check (send)"
and if using regex - Layer7 timeout, info: " at step 3 of tcp-check (expect regex)"
```
Also i compared some captures with tcpdump and as i can see from them, haproxy is sending data, and packets are coming back with a valid response but for an unknown reason haproxy is throwing 'layer7 timeout'

I've also tried to send and expect binary, but the problem remains the same.

## Do you have any idea what may have caused this?
no
## Do you have an idea how to solve the issue?
no
## What is your configuration?

```
frontend FT__memcached
	bind :11211
	mode tcp
	option tcplog
	timeout client 1m
	default_backend BK__memcached

backend BK__memcached
	mode tcp
	timeout server 10s
	timeout connect 10s
	option tcp-check
	tcp-check connect port 11212
	tcp-check send stats\r\n
	tcp-check expect string STAT\ pid
#	tcp-check expect rstring ^STAT\ pid\ (.*[\d]).*$
	tcp-check send quit\r\n
	balance source
	default-server check inter 2000 fall 3 on-marked-down shutdown-sessions
	server srv01 192.168.24.241:11212
	server srv02 192.168.24.242:11212
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.3.2-d522db7 2020/11/28 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2022.
Known bugs: http://www.haproxy.org/bugs/bugs-2.3.2.html
Running on: Linux 5.10.3-arch1-1 #1 SMP PREEMPT Sun, 27 Dec 2020 10:50:46 +0000 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -march=x86-64 -mtune=generic -O2 -pipe -fno-plt
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=4).
Built with OpenSSL version : OpenSSL 1.1.1h  22 Sep 2020
Running on OpenSSL version : OpenSSL 1.1.1i  8 Dec 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.4.1
Built with network namespace support.
Built with the Prometheus exporter as a service
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.44 2020-02-12
Running on PCRE version : 8.44 2020-02-12
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace

```
```
Linux srv01 5.10.3-arch1-1 #1 SMP PREEMPT Sun, 27 Dec 2020 10:50:46 +0000 x86_64 GNU/Linux
```

## Additional information (if helpful)
OS archlinux

**Comment by capflam** [comment]
I confirm the issue. But I have the same error with the string or the regex matching, on the 4th tcp-check rule. BTW, I will provide a patch soon.

**Comment by capflam** [comment]
FYI, I pushed a fix to the 2.4-dev. Thanks !
